internal/util/usage/usage.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package usage
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	kyaml "sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	CNRMMetricsAnnotation              = "cnrm.cloud.google.com/blueprint"
+	DisableKptUsageTrackingEnvVariable = "KPT_DISABLE_USAGE_TRACKING"
+)
+
+// Tracker is used to track the usage the of kpt
+type Tracker struct {
+	// PackagePaths is the package paths to be tracked for usage
+	PackagePaths []string
+
+	Resources []*kyaml.RNode
+
+	cmdGroup string
+}
+
+// TrackAction invokes Tracker kyaml filter on the resources in input packages paths
+func (t *Tracker) TrackAction(cmdGroup string) {
+	// users can opt-out by setting the "KPT_DISABLE_USAGE_TRACKING" environment variable
+	if os.Getenv(DisableKptUsageTrackingEnvVariable) != "" {
```

Comment:
not sure if people will like the disable version of it.  There might need to be a build for GCP that's different than anyone else.  I think generally people want to know if they are sharing usage metrics even though outside of the GCP environment this goes to nowhere.

Reply:
Couldn't parse this. Can you say more Mike ?

Reply:
sorry, let me try again.  I would say in a CLI tool the metrics being sent anywhere should be an opt in rather an an opt out.  So if the user just gets the tool and downloads it the tool should err on the side of not even attempting to send the information anywhere.

Reply:
One thing I would like to highlight is that the instrumentation in this case is very different from the traditional tracking where command line program actively talks to a remote server and report the usage. In our case, we are just annotating (or augmenting) the resources and there is no active tracking involved. Once the resources are submitted to a kubernetes control plane, the control-plane owner can make use of this attribution. Users can leverage this attribution annotation for their own analysis also. In case of KCC resources running managed through Config Controller, we plan to have our own pipeline to process this attribution for KCC resources. So, this is very different from traditional tracking, so default `opt-in` seems reasonable.
Or may be `tracking` is a wrong term to use here ? we can use `attribution` term ? `KPT_DISABLE_ATTRIBUTION` ?

Reply:
Yes. The name is confusing, we are not sending heart-beats to track metrics. We just add annotation to the resources. So this should be opt-out IMO. I am changing the name to `KPT_DISABLE_ATTRIBUTION`.